### PR TITLE
feat: migrate to vue 3

### LINKED ISSUE #203: Vue 3 compatibility
**Describe the bug**
I started a project in vue3 but can't get this plugin working in it.
plugin version: 2.3.13

**Screenshots**

**Desktop:**
 - OS: Manjaro Linux
 - Browser: Chrome
 - Version: 86.0.4240.75

**Smartphone:**
 - Device: Xiaomi Mi A3
 - OS: Android 10
 - Browser: Chrome
 - Version: 85.0.4183.127

### PR DESCRIPTION
**Pull request by dargmuesli**
Resolves #203.

Uses Typescript, Vite and Vitepress.
Uses ~`@undecaf/barcode-detector-polyfill`~ `@sec-ant/barcode-detector`

### CODE REVIEW

**Comment by gruhn** [comment]
Really cool thanks. Is the demo page deployed somewhere, so I can test it?

**Comment by dargmuesli** [comment]
Now it is: https://dargmuesli.github.io/vue-qrcode-reader/ :partying_face:

**Comment by gruhn** [comment]
I also noticed that the "visual tracking" works slightly different on the new demo page. Essentially "outline" works like "bounding box" now:

Screenshot 2023-06-01 at 13 39 50

Screenshot 2023-06-01 at 13 40 29

**Comment by gruhn** [comment]
But all in all, really cool so far. Vitepress really looks much nicer. And also dark mode for free 👍

**Comment by dargmuesli** [comment]
Thank you for the feedback! You got all the little things I was not sure about :wink: 
I will dive deeper later :)

File: src/misc/shimGetUserMedia.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import adapter from 'webrtc-adapter'
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/misc/shimGetUserMedia.ts:1]: Gotcha, problem being that `webrtc-adapter` does not export anything subpaths currently, which leads to typescript issues. Would need a PR at https://github.com/webrtcHacks/adapter I think.

File: src/misc/camera.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  //   videoEl.value.mozSrcObject = stream
+  // } else if (window.URL.createObjectURL) {
+  //   // @ts-ignore
+  //   videoEl.value.src = window.URL.createObjectURL(stream)
+  // } else if (window.webkitURL) {
+  //   // @ts-ignore
+  //   videoEl.value.src = window.webkitURL.createObjectURL(stream)
+  // } else {
+  //   videoEl.value.src = stream.id
+  // }
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/misc/camera.ts:133]: Typescript wasn't happy about those statements. Some googling showed that `mozSrcObject` does not exist anymore, for example, so I removed it for now. Would need to try edge case browsers, but for a new major release (dropping Vue 2 support e.g.) that could be added later, I thought.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// // Auto-install
+// let GlobalVue = null
+// if (typeof window !== 'undefined') {
+//   GlobalVue = window.Vue
+// } else if (typeof global !== 'undefined') {
+//   GlobalVue = global.Vue
+// }
+// if (GlobalVue) {
+//   GlobalVue.use(plugin)
+// }
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/index.ts:31]: I suspected those statements to always bundle vue-qrcode-reader, but I'm new to vue plugin auto imports, so I'm not sure how they work rn. Also, as you might expect, typescript complained :wink:

**Comment by dargmuesli** [comment]
Not sure why the outline now behaves like a bounding box now :joy: haven't looked into this implementation yet, but will do so later, as promised :)

File: src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (gruhn) [reply src/index.ts:31]: I mean I didn’t came up with this. This was part of the template plugin, this project is based on. Could very well be that the best practice has changed.

I think the point of this is for non-npm users. So if you include the package just via script tag, then it’s „installed“ immediately. Probably not a big deal, since we are introducing breaking changes anyway. But then the install instructions in the README should probably be adjusted.

File: src/misc/shimGetUserMedia.js [hunk 4]
```diff
@@ -1,29 +0,0 @@
-import { shimGetUserMedia as chromeShim } from "webrtc-adapter/src/js/chrome/getusermedia";
-import { shimGetUserMedia as edgeShim } from "webrtc-adapter/src/js/edge/getusermedia";
-import { shimGetUserMedia as firefoxShim } from "webrtc-adapter/src/js/firefox/getusermedia";
-import { shimGetUserMedia as safariShim } from "webrtc-adapter/src/js/safari/safari_shim";
-import { detectBrowser } from "webrtc-adapter/src/js/utils";
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/misc/shimGetUserMedia.js:5]: Reverted those changes

File: src/misc/shimGetUserMedia.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/misc/shimGetUserMedia.ts]: Reverted those changes

File: src/misc/camera.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/misc/camera.ts:133]: Reverted those changes

File: src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/index.ts]: Reverted those changes

File: docs/.vitepress/components/demos/Fullscreen.vue [hunk 5]
```diff
@@ -2,7 +2,7 @@
   <div :class="{ 'fullscreen': fullscreen }" ref="wrapper" @fullscreenchange="onFullscreenChange">
     <qrcode-stream @init="logErrors">
       <button @click="fullscreen = !fullscreen" class="fullscreen-button">
-        <img :src="$withBase(fullscreenIcon)" alt="toggle fullscreen" />
+        <img :src="fullscreenIcon" alt="toggle fullscreen" />
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply docs/.vitepress/components/demos/Fullscreen.vue]: Readded `withBase`

File: docs/.vitepress/components/demos/Torch.vue [hunk 6]
```diff
@@ -6,7 +6,7 @@
 
     <qrcode-stream :torch="torchActive" @init="onInit">
       <button @click="torchActive = !torchActive" :disabled="torchNotSupported">
-        <img :src="$withBase(icon)" alt="toggle torch">
+        <img :src="icon" alt="toggle torch">
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply docs/.vitepress/components/demos/Torch.vue]: Readded `withBase`

File: docs/.vitepress/components/demos/SwitchCamera.vue [hunk 7]
```diff
@@ -10,7 +10,7 @@
 
     <qrcode-stream :camera="camera" @init="onInit">
       <button @click="switchCamera">
-        <img :src="$withBase('/camera-switch.svg')" alt="switch camera">
+        <img src="/camera-switch.svg" alt="switch camera">
       </button>
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply docs/.vitepress/components/demos/SwitchCamera.vue:10]: Readded `withBase`

File: docs/.vitepress/components/demos/ScanSameQrcodeMoreThanOnce.vue [hunk 8]
```diff
@@ -5,7 +5,7 @@
 
     <qrcode-stream :camera="camera" @decode="onDecode" @init="onInit">
       <div v-show="showScanConfirmation" class="scan-confirmation">
-        <img :src="$withBase('/checkmark.svg')" alt="Checkmark" width="128px" />
+        <img src="/checkmark.svg" alt="Checkmark" width="128px" />
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply docs/.vitepress/components/demos/ScanSameQrcodeMoreThanOnce.vue]: Readded `withBase`

**Comment by dargmuesli** [comment]
Now that `barcode-reader` is used again, the outline is shown correctly again. I kept `@undecaf/barcode-detector-polyfill` only as devdependency to provide types for typescript.

File: src/components/QrcodeStream.vue [hunk 9]
```diff
@@ -6,279 +6,283 @@
...
-    clearCanvas(canvas) {
-      const ctx = canvas.getContext("2d");
-      ctx.clearRect(0, 0, canvas.width, canvas.height);
+      props.track(adjustedCodes, ctx)
+    } else {
+      clearCanvas(canvas)
     }
-
   }
-};
```
Comment (gruhn) [inline src/components/QrcodeStream.vue:281]: Ok, there are so many changes in this file. It's hard for me to keep track of what actually has changed. Especially the `QrcodeStream.vue` is a bit sensitive to changes. I'm sure it's a good idea to migrate to Composition API. But I would be grateful if this can be done in a second step.

File: docs/.vitepress/config.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  description: 'A set of Vue.js components for detecting and decoding QR codes.',
+  lang: 'en-US',
+  lastUpdated: true,
+  themeConfig: {
+    editLink: {
+      pattern: 'https://github.com/gruhn/vue-qrcode-reader/edit/main/docs/:path'
+    },
+    footer: {
+      message: 'Released under the MIT License.',
+      copyright: 'Copyright © 2019-present Jonas Thelemann'
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply docs/.vitepress/config.ts:14]: Whoops, copy-paste-error, embarrassing. Removed!

File: src/components/QrcodeStream.vue (hunk 9 above)
Comment (dargmuesli) [reply src/components/QrcodeStream.vue:281]: :0 I worked on that based on that you said

> Only acceptance criterium: I can view and test the deployed demo page without doing any setup on my machine.

https://github.com/gruhn/vue-qrcode-reader/issues/203#issuecomment-1431324053

Now I put much effort into this and also typescript being included now really makes it hard to make mistakes.

File: docs/.vitepress/components/demos/DragDrop.vue [hunk 11]
```diff
@@ -30,28 +29,28 @@ export default {
...
 
         this.result = content
         this.error = null
       } catch (error) {
         if (error.name === 'DropImageFetchError') {
-          this.error = 'Sorry, you can\'t load cross-origin images :/'
+          this.error = "Sorry, you can't load cross-origin images :/"
         } else if (error.name === 'DropImageDecodeError') {
-          this.error = 'Ok, that\'s not an image. That can\'t be decoded.'
+          this.error = "Ok, that's not an image. That can't be decoded."
```
[the remark this answers is not in the document]
Comment (gruhn) [reply docs/.vitepress/components/demos/DragDrop.vue:42]: alright, if it's standard then let's go with it 👍

File: src/components/QrcodeStream.vue (hunk 9 above)
Comment (gruhn) [reply src/components/QrcodeStream.vue:281]: I know, sorry. Typescript is great but not a silver bullet. The thing is `QrcodeStream` is very fragile. A lot of stuff is happening async and the order is important. It's also particularly easy to break responsiveness. Ok, but are you saying you really changed nothing regarding the components function?

File: src/components/QrcodeStream.vue [hunk 12]
```diff
@@ -6,279 +6,283 @@
...
-      class="qrcode-stream-camera"
-    ></canvas>
+    <canvas ref="pauseFrameRef" v-show="!shouldScan" class="qrcode-stream-camera" />
 
-    <canvas ref="trackingLayer" class="qrcode-stream-overlay"></canvas>
+    <canvas ref="trackingLayerRef" class="qrcode-stream-overlay" />
 
     <div class="qrcode-stream-overlay">
-      <slot></slot>
+      <slot />
```
Comment (gruhn) [inline src/components/QrcodeStream.vue:22]: For example this. Are these two definitely the same?

Comment (dargmuesli) [reply src/components/QrcodeStream.vue:22]: Yes

File: src/components/QrcodeStream.vue (hunk 9 above)
Comment (dargmuesli) [reply src/components/QrcodeStream.vue:281]: I think all documented demos work as expected, so I don't know what else I could ensure. Did you find any issues?

Comment (dargmuesli) [reply src/components/QrcodeStream.vue:281]: I mean I also get why there is a higher likelyhood that some changes break something here, but pulling it out here and creating another PR for it would have us looking at the same diff, wouldn't it?

**Comment by gruhn** [comment]
Another difference I noticed is the "Scan Same QR Code More Than Once" demo. On detection, it's supposed to pause the stream and show a transparent-white overlay with a green checkmark in the center:

https://gruhn.github.io/vue-qrcode-reader/demos/ScanSameQrcodeMoreThanOnce.html

But here container just seems to collapse to zero height:

https://dargmuesli.github.io/vue-qrcode-reader/demos/ScanSameQrcodeMoreThanOnce.html

File: src/misc/scanner.ts [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+ */
+export const keepScanning = async (
+  videoElement: Ref<HTMLVideoElement | undefined>,
+  {
+    detectHandler,
+    locateHandler,
+    minDelay
+  }: { detectHandler: any; locateHandler: any; minDelay: any }
+) => {
+  const barcodeDetector = new (await import('barcode-detector')).default({ formats: ['qr_code'] })
```
Comment (gruhn) [inline src/misc/scanner.ts:51]: Why do you use a dynamic import here?

File: docs/.vitepress/components/demos/CustomTracking.vue [hunk 14]
```diff
@@ -4,28 +4,27 @@
...
       <select v-model="selected">
         <option v-for="option in options" :key="option.text" :value="option">
-          {{ option.text }}
+          {{ option.text }}
         </option>
       </select>
     </p>
 
-    <qrcode-stream :key="_uid" :track="selected.value" @init="logErrors" />
+    <qrcode-stream :track="selected.value" @init="logErrors" />
```
Comment (gruhn) [inline docs/.vitepress/components/demos/CustomTracking.vue:12]: I don't quite remember what this was good for. But why did you remove it?

File: src/misc/scanner.ts (hunk 13 above)
Comment (dargmuesli) [reply src/misc/scanner.ts:51]: To not pollute `global` and to save on bundle size. But maybe that's necessary for `<script>` inclusion? I never added dependencies without a package manager, so could be that you know more about this? :pray:

File: docs/.vitepress/components/demos/CustomTracking.vue (hunk 14 above)
Comment (dargmuesli) [reply docs/.vitepress/components/demos/CustomTracking.vue:12]: Because it "[is reserved for internal use and it's important to keep it private (and not rely on it in user code) so that we keep the flexibility to change its behavior for potential future use cases](https://github.com/vuejs/vue/issues/5886#issuecomment-308625735)".

**Comment by dargmuesli** [comment]
[ScanSameQrcodeMoreThanOnce.html](https://dargmuesli.github.io/vue-qrcode-reader/demos/ScanSameQrcodeMoreThanOnce.html) fixed! :raised_hands:

File: src/misc/scanner.ts (hunk 13 above)
Comment (gruhn) [reply src/misc/scanner.ts:51]: I would argue polyfills/shims are the exception, because their purpose is specifically to patch `global` and “pretend” that you are running in a more modern browser. That being said, this polyfill is not modifying `global` on import. So importing it the normal way is not polluting. 

Some other things to consider:

* is there a performance penalty for re-importing before every scan? The browser probably caches the asset but still.
* If we only import right before scanning there might be some awkward delay where the camera is loaded and ready but it takes some more time before the scanning actually works

File: docs/.vitepress/components/demos/CustomTracking.vue (hunk 14 above)
Comment (gruhn) [reply docs/.vitepress/components/demos/CustomTracking.vue:12]: Alright 👍 but can you guess what it’s purpose might have been? I think it had something to do with rendering

Comment (dargmuesli) [reply docs/.vitepress/components/demos/CustomTracking.vue:12]: Normally you'd like to use keys for lists, but I don't see a list being used here. So I can't guess why it was used here.

File: src/misc/scanner.ts (hunk 13 above)
Comment (dargmuesli) [reply src/misc/scanner.ts]: Reverting that showed that the change in https://github.com/gruhn/barcode-detector/pull/8 is necessary. I'll update this PR once the other PR is merged :)

Comment (dargmuesli) [reply src/misc/scanner.ts]: As @Sec-ant [pointed out](https://github.com/gruhn/barcode-detector/issues/6#issuecomment-1588414162) I was able to make use of `@sec-ant/barcode-detector` now. Let me know what you think! :)

Comment (dargmuesli) [reply src/misc/scanner.ts]: Eventhough I tested it locally before, using barcode-detector v1.0.2 still results in an error `unexpected token 'export'`: https://github.com/dargmuesli/vue-qrcode-reader/actions/runs/5251319091/jobs/9486148441#step:6:174 :thinking: if the previously mentioned sec-ant issue got fixed, would it be fine to move to that library?

Comment (dargmuesli) [reply src/misc/scanner.ts]: The mentioned issue is fixed now.

**Comment by dargmuesli** [comment]
@gruhn what do you think about my changes? :)

**Comment by gruhn** [comment]
Haven’t had time to further look into it deeply, sorry

File: src/components/QrcodeStream.vue [hunk 15]
```diff
@@ -6,279 +6,282 @@
...
+const isMounted = ref(false)
+
+// computations
+const shouldScan = computed(
+  () =>
+    isMounted.value &&
+    props.camera !== 'off' &&
+    cameraInstance.value &&
+    videoRef.value &&
+    videoRef.value.readyState > 1
```
Comment (gruhn) [inline src/components/QrcodeStream.vue:72]: I think technically we don't have to check for 
```
videoRef.value &&
videoRef.value.readyState > 1
```
here but correct me if I'm wrong. `videoRef.value` is truthy iff `isMounted.value` is `true` and `videoRef.value.readyState > 1` holds iff `cameraInstance.value` is truthy. The latter holds because we [await the `loadeddata` event](https://github.com/gruhn/vue-qrcode-reader/blob/edf2635c2e424d3bb443776c9f1fbc6e3440d6d3/src/misc/camera.ts#L135) when we initialize `cameraInstance`. Also, I'm not sure about `videoRef.value` but I guess at least `videoRef.value.readyState` is not reactive so Vue can't detect when this value changes, right?

File: src/misc/camera.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (gruhn) [reply src/misc/camera.ts:133]: > mozSrcObject does not exist anymore

Right, but the idea is to support legacy browsers. 

> When I comment the disabled code back in, I get an error in Firefox and Chrome.

What kind of error?

[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/misc/camera.ts:133]: Chrome: `DOMException: Could not start video source`
Firefox: `TypeError: URL.createObjectURL: Argument 1 is not valid for any of the 1-argument overloads.` (at `camera.ts:127`)

File: .github/workflows/main.yml
[the remark this answers is not in the document]
Comment (dargmuesli) [reply .github/workflows/main.yml:1]: I have not tried that as I prefer to use GitHub's action to upload to their pages instead of 3rd party actions, which I consider unnecessary by now.

File: .github/workflows/deploy.yml [hunk 16]
```diff
@@ -0,0 +1,36 @@
+name: Deploy
+
+on:
+  workflow_dispatch: {}
+  push:
+    branches:
+      - vue-3
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply .github/workflows/deploy.yml:8]: Added :heavy_check_mark:

File: docs/public/debug-memory-leak.html
[the remark this answers is not in the document]
Comment (dargmuesli) [reply docs/public/debug-memory-leak.html:1]: Changed :heavy_check_mark:

File: src/components/QrcodeStream.vue (hunk 15 above)
Comment (dargmuesli) [reply src/components/QrcodeStream.vue]: According to my testing, you are right!

File: src/misc/camera.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (gruhn) [reply src/misc/camera.ts:133]: Hmm weird :thinking: can you reproduce this on the current demo page?

[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/misc/camera.ts:133]: Do you mean I should push the commented-in code to the demo page so you can check as well? The mentioned code is currently not part of the deployment.

[the remark this answers is not in the document]
Comment (gruhn) [reply src/misc/camera.ts:133]: No I mean the master branch demo page:

https://gruhn.github.io/vue-qrcode-reader/

If the same issue occurs there, I guess it’s really a problem with this code in particular. Otherwise it must be some interaction with other changes. And that might be worrisome.

[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/misc/camera.ts:133]: I do not have this problem on the current demo page, you're right. Maybe it's related to the new bundling. I'll need to check, but am currently on vacation so it might take a fair bit of time until I get around to that :pray:

[the remark this answers is not in the document]
Comment (gruhn) [reply src/misc/camera.ts:133]: Alright, other than that I’m done complaining 😅

[the remark this answers is not in the document]
Comment (gruhn) [reply src/misc/camera.ts:133]: Nevermind, I think I found the issue. Before it was:
```
if (videoEl.srcObject !== undefined)
```
which is supposed to check whether the property is defined in the browser at all. If `srcObject` is defined but empty, it's value is `null`. You had
```
if (videoEl.srcObject)
```
and that's guaranteed to fail.

**Comment by gruhn** [comment]
Alright, I'm gonna merge this bad boy. Thanks for all the work!
